src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+                                        foreachCollectionExpression)))));
+
+                collectionStatement =
+                    collectionStatement.ReplaceToken(collectionStatement.GetFirstToken(), collectionStatement.GetFirstToken().WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+
+                editor.InsertBefore(foreachStatement, collectionStatement);
+            }
+
+            // create new index varialbe name
+            var indexString = CreateUniqueMethodName(model, foreachStatement.Statement, "i");
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(foreachInfo, parseOption, collectionVariableName, indexString);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(SyntaxFactory.IdentifierName("var"),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                        SyntaxFactory.Identifier(indexString).WithAdditionalAnnotations(RenameAnnotation.Create()),
+                        default,
+                        SyntaxFactory.EqualsValueClause(
+                            SyntaxFactory.LiteralExpression(SyntaxKind.NumericLiteralExpression, SyntaxFactory.Literal(0)))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                SyntaxFactory.BinaryExpression(SyntaxKind.LessThanExpression, SyntaxFactory.IdentifierName(indexString), SyntaxFactory.ParseExpression($"{collectionVariableName}.{foreachInfo.CountName}")),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(SyntaxFactory.PostfixUnaryExpression(SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexString))),
+                bodyStatement);
+
+            // let leading and trailing trivia set
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                forStatement = forStatement.ReplaceToken(forStatement.ForKeyword, forStatement.ForKeyword.WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+            }
+
+            forStatement = forStatement.ReplaceToken(forStatement.CloseParenToken, forStatement.CloseParenToken.WithTrailingTrivia(foreachStatement.CloseParenToken.TrailingTrivia));
+
+            editor.ReplaceNode(foreachStatement, forStatement);
+        }
+
+        private ExpressionSyntax GetForeachCollection(SyntaxNode node)
+        {
+            var foreachStatement = (ForEachStatementSyntax)node;
+
+            // use original text, not value text
+            return foreachStatement.Expression;
+        }
+
+        private StatementSyntax GetForLoopBody(ForeachInfo foreachInfo, ParseOptions parseOption, string collectionVariableName, string indexString)
+        {
+            var foreachStatement = (ForEachStatementSyntax)foreachInfo.ForeachStatement;
```

Comment:
what about ForEachVariableStatementSyntax?

Reply:
not supported yet. will do vb first and then for to foreach and then might come back for tuple.